ENH: ndimage: uarray based backend support

AnirudhDagar opened the pull request:
#### Reference issue
See gh-14353

#### What does this implement/fix?
This adds support for [`uarray`](https://uarray.org/en/latest/index.html) backend implementation to `scipy.ndimage`.

A simple example tested after adding dummy CuPy backend on [this branch](https://github.com/AnirudhDagar/scipy/blob/ff8143535d7f0b1bfd09be6f8ce145c7b6b896ed/scipy/ndimage/backends/cupy_backend.py#L64), verifies the backend.

```python
x_cu, x_nu = cp.arange(10), np.arange(10)

y_scipy = ndimage.correlate1d(x_nu, weights=np.array([1, 2.5]))

with scipy.ndimage.set_backend(_cupy_backend):
    y_cupy = ndimage.correlate1d(x_cu, cp.array([1, 2.5]))

print((y_cupy, type(y_cupy)), (y_scipy, type(y_scipy)) , sep='\n')

>>> (array([ 0,  2,  6,  9, 13, 16, 20, 23, 27, 30]), <class 'cupy._core.core.ndarray'>)
(array([ 0,  2,  6,  9, 13, 16, 20, 23, 27, 30]), <class 'numpy.ndarray'>)
```

#### Question
**EDIT 3**:  This has been resolved, and we now run the whole `ndimage` test suite based on `uarray`.
~Currently, I've not added any additional tests. We're now running the existing test suite via the multimethods.~
Edit 2: We now have separate tests for testing the backend based on the approach below.

~We could do something similar to [test_backend.py](https://github.com/scipy/scipy/blob/master/scipy/fft/tests/test_backend.py) and [mock_backend.py](https://github.com/scipy/scipy/blob/master/scipy/fft/tests/mock_backend.py) just to make sure that a third party backend can be properly registered and used via a context manager, but adding every function under ndimage in thoses tests is not something we should be doing (it is redundant to the current test suite since we set the `scipy` global backend anyway and will probably be a repetition).~

Edit 1: I realized it doesn't really matter since the only thing that is being tested is whether dispatching works through the mock backend or not.

~Let me know if it is necessary to include this type of test for `ndimage` as well.~

@rgommers just to confirm, do you want me to address the private/public namespace concerns of `ndimage` from cupy/cupy#5047 in this PR itself? Or should I do that subsequently, once this PR lands to keep them separated?

cc @peterbell10 @grlee77 

#### Additional Information
I've also removed the use of unnecessary private method `scipy.ndimage.filters.rank_filter` in favour of `scipy.ndimage.rank_filter` from the test suite.

AnirudhDagar committed:
ENH: Create _api.py for importing original methods

AnirudhDagar committed:
ENH: Setup ndimage uarray backend

DOC: set backend str to scipy and fix docs

DOC: Fix notes about uarray domain hierarchy

MAINT: explicitly define ndimage __all__

AnirudhDagar committed:
ENH: Add ndimage multimethods

 * ENH: Add filters multimethods

 * ENH: Add fourier multimethods

 * ENH: Add interpolation, measurements, morphology multimethods

rgommers commented:
> @rgommers just to confirm, do you want me to address the private/public namespace concerns of `ndimage` from [cupy/cupy#5047](https://github.com/cupy/cupy/pull/5047) in this PR itself? Or should I do that subsequently, once this PR lands to keep them separated?

Separate is better, since it's an orthogonal change. Let me open an issue about that with some motivation and history, so we have a single place that we can reference when we start fixing this in various submodules.

AnirudhDagar commented:
Yeah, that sounds good! Will it pick it up after you raise the issue. Thanks:)

rgommers commented:
> Currently, I've not added any additional tests. We're now running the existing test suite via the multimethods.
> 
> We could do something similar to [test_backend.py](https://github.com/scipy/scipy/blob/master/scipy/fft/tests/test_backend.py) and [mock_backend.py](https://github.com/scipy/scipy/blob/master/scipy/fft/tests/mock_backend.py) just to make sure that a third party backend can be properly registered and used via a context manager, but adding every function under ndimage in thoses tests is not something we should be doing

A separate test to ensure a third-party backend works but not running duplicate tests sounds very reasonable to me. Let's wait for @peterbell10 or @grlee77 to comment though, they know more about how testing in `scipy.fft` works.

grlee77 commented:
> A separate test to ensure a third-party backend works but not running duplicate tests sounds very reasonable to me. Let's wait for @peterbell10 or @grlee77 to comment though,

We could copy the mock backend approach from the fft tests. I don't know that it is necesarry to mock every function in the ndimage namespace, though.

peterbell10 commented:
>I don't know that it is necesarry to mock every function in the ndimage namespace, though.

I think it's valuable to test that each function really does do dispatching. If you're just worried about lines of code, then you could always define a backend that just records the arguments to `__ua_function__ ` so you wouldn't need any code to mock individual functions. Something like:
```python
class MockBackend:
    __ua_domain__ = "numpy.scipy"

    def __init__(self, return_value=None):
        self.num_calls = 0
        self.last_args = None
        self.return_value = return_value

    def __ua_function__(self, method, args, kwargs):
        self.num_calls += 1
        self.last_args = (method, args, kwargs)
        return self.return_value
```

grlee77 commented on scipy/ndimage/tests/test_filters.py:
```diff
@@ -11,7 +11,7 @@
 
 from scipy import ndimage
-from scipy.ndimage.filters import _gaussian_kernel1d, rank_filter
+from scipy.ndimage.filters import _gaussian_kernel1d
```
Ralf had [indicated that `scipy.ndimage.filters`](https://github.com/cupy/cupy/pull/5047#issuecomment-814914710) etc. are not intended to be public. All functions should be imported from `scipy.ndimage`. Requiring importing from the top level for to get the multimethod versions would seem in line with this.

In practice, I am sure there are many instances of downstream libraries not following this convention of only importing from `scipy.ndimage` itself (we had a few that I recently fixed in `scikit-image`). It probably doesn't help that `scikit-image` does not import very many things at the top level and requires explicitly importing from subpackages like `filters` or `morphology` instead!

rgommers replied:
> Ralf had [indicated that `scipy.ndimage.filters`](https://github.com/cupy/cupy/pull/5047#issuecomment-814914710) etc. are not intended to be public. All functions should be imported from `scipy.ndimage`.

Just filed an issue for that: gh-14360

peterbell10 replied:
I think when the same question came up for `fftpack`, someone linked to something like a "scipy cookbook" that imported from submodules. So, AFAIK these are assumed to be public API; intentional or not.

My opinion is either break `import scipy.ndimage.fiters` completely, or make it export the multimethods.

AnirudhDagar replied:
If we make `scipy.ndimage.<module>` export the multimethods, we are actually giving more incentive and motivation for the new user to use these private modules. Instead, we can add a note in the docs as `Do not use this private namespace, instead import like so...` and deprecate these private modules which are public looking (without underscore ones) gradually.

peterbell10 replied:
A documentation comment on a particular function isn't really enough IMO. Existing code will lie incorrect with nobody noticing and anyone who copies from existing code will perpetuate the discouraged usage. It needs to actually emit a warning or error.

>If we make `scipy.ndimage.<module>` export the multimethods, we are actually giving more incentive and motivation for the new user to use these private modules.

Having `scipy.ndimage.<module>.<func>`  behave differently than `scipy.ndimage.<func>` is a really bad idea IMO.

AnirudhDagar replied:
I agree with emitting a warning in such a case. This is actually linked to gh-14360 anyway which proposes similar way and needs more discussion. It is probably not a good idea to suddenly break the use of private modules though.

> Having scipy.ndimage.<module>.<func> behave differently than scipy.ndimage.<func> is a really bad idea IMO.

I don't have any strong opinions here. @rgommers @grlee77 what do you think?

AnirudhDagar replied:
gentle ping @rgommers.

grlee77 replied:
What do you think of the following approach?
   1.) `git move filters.py _filters.py` 
   2.) modify `_api.py` to import from `_filters` instead of `filters`
   3.) Make a new `filters.py` with a `__getattr__` similar to [Peter's suggestion here](https://github.com/scipy/scipy/pull/14419/files#r672433228), to call the multimethod from `_multimethods.py`. (The decision on adding a deprecation warning can be left to a separate PR)
   4.) repeat steps 1-3 for `morphology.py` and other subpackages

The suggestion above seems like it would require minimal refactoring in this PR, but still allow the multimethods to work if imported from subpackages.

AnirudhDagar committed:
ENH: Fix multimethods with positional args input

AnirudhDagar committed:
STY: PEP8 Formatting

AnirudhDagar commented on scipy/ndimage/_multimethods.py:
```diff
@@ -0,0 +1,756 @@
+        if "footprint" in kw_out:
+            kw_out["footprint"] = dispatchables[1]
+        if "output" in kw_out:
+            kw_out["output"] = dispatchables[2]
```
I've tried to reuse arg replacers wherever possible and have written an argument replacer for each unique signature. This should address the issue of positional arguments.

AnirudhDagar commented:
> I think it's valuable to test that each function really does do dispatching. If you're just worried about lines of code, then you could always define a backend that just records the arguments to __ua_function__ so you wouldn't need any code to mock individual functions. Something like:

@peterbell10 could you please explain this a bit more for me. I'm not very sure how we can call different functions without passing the required arguments which unlike `scipy.fft` will be more than just a single input `x` to test all the functions. Also, I don't understand why the `__ua_function__` in your code snippet has no call to the implemented functions.
If you want, I can go ahead and add all the functions with separate `test_backend_call_<unique_signature>` tests for different functions. 

Thanks in advance :)

AnirudhDagar reviewed (commented):
I've followed a specific pattern for the arg replacer names based on reusability. See below.

AnirudhDagar committed:
TST: add ndimage uarray backend tests

peterbell10 commented on scipy/ndimage/_multimethods.py:
```diff
@@ -0,0 +1,954 @@
+    """
+    uarray argument replacer for fourier multimethods.
+    """
+    def self_method(input, arg2, n=-1, axis=-1, output=None, *args, **kwargs):
```
> `arg2`

This doesn't work, the names need to match in order for keyword-arguments to work correctly.

AnirudhDagar replied:
Thanks, Peter for catching this. Sorry I missed this simple thing when making the changes for `arg` inputs issue.

AnirudhDagar committed:
ENH: Fix multimethods called kwarg like input for a positional arg

AnirudhDagar committed:
STY: Follow convention for private funcs

AnirudhDagar committed:
MAINT: Add __all__ to _backend and _multimethods

AnirudhDagar commented on scipy/ndimage/_multimethods.py:
```diff
@@ -0,0 +1,954 @@
+from scipy.ndimage import _api
+
+
+mark_non_coercible = lambda x: Dispatchable(x, np.ndarray, coercible=False)
```
Done! :)

AnirudhDagar reviewed (commented):
Thanks for the careful review, I've made the changes.

AnirudhDagar committed:
MAINT: Use uarray multimethods for ndimage.<module>.func call

AnirudhDagar commented:
I've made the changes according to @grlee77's [comment](https://github.com/scipy/scipy/pull/14356#discussion_r672534127) which sounds like a good approach:

1. This way we can keep the `uarray` backend through the multimethods, and the actual SciPy implementation of the functions **separate**. This offers better maintainability.
2. Also, with this change, the behaviour between `ndimage.<module>.<func>` and `ndimage.<func>` remains the same, both calling the multimethods, addressing concerns raised [here](https://github.com/scipy/scipy/pull/14356#discussion_r668071635) by @peterbell10.

A couple of other things to take note of:
* I've made some changes to the tests calling private functions. They need to be called from the new (renamed) private module namespace since these private functions are not available in the public namespace. 
* Also, in anticipation of the deprecation warnings (to be added in a subsequent PR, similar to #14419 ), I've removed any instance of function call through the `ndimage.<module>` namespace.

I guess some of the changes here overlap with the changes to be made for the deprecation of public like modules in ndimage but that's okay. I'll add a link to this comment in the description of the deprecation PR, just in case someone wants to understand the changes made, later.

AnirudhDagar commented:
Ahh, need to fix the merge conflicts before `git mv` otherwise git forgets that the files are just being renamed.

AnirudhDagar committed:
BUG: Fix rotate arg replacer

AnirudhDagar commented:
@peterbell10 @grlee77 @rgommers Thank you for your patience with this one.

As I mentioned earlier, the diff is extremely large when following @grlee77 suggestion in a single PR.

I propose to split this PR into three sequential PRs each adding the following:

1. renaming modules (from `module.py` to `_module.py`)
2. `uarray multimethods` (adds `uarray` support as well as brings back `module.py` for backwards compatibility)
3. deprecating `ndimage` public looking private namespace (i.e. ndimage.module.func will raise warning)

To make sure the proposed method works in mitigating the huge diff, I tried it out on my own fork.
Please refer to the following in order to understand the proposed approach completely.

1. anirudhdagar/scipy#1 (explains the issue of huge diff)
2. anirudhdagar/scipy#3 (demonstrates step 1, i.e renaming modules)
3. anirudhdagar/scipy#4 (demonstrates step 2, i.e `uarray multimethods` )
4. anirudhdagar/scipy#5 (demonstrates step 2, i.e deprecating `ndimage` modules)

This will keep git blame (You can look at my [`master_copy`](https://github.com/AnirudhDagar/scipy/tree/master_copy/scipy/ndimage) branch) and also avoid the huge diff change as seen in this PR currently.

Please let me know your opinion on this, if it looks good, then I'll go ahead :))

ev-br commented:
Drive-by comment: you can always send a series of PRs where PR N depends on N-1, and rebase N on master once N-1 is merged.

rgommers commented:
> * renaming modules (from `module.py` to `_module.py`)
> * `uarray multimethods` (adds `uarray` support as well as brings back `module.py` for backwards compatibility) (**Edit 1**: we can actually use this PR after minor refactoring for this purpose)

This isn't ideal, because we cannot merge PR 1 separately - that will break master, and the second PR may need more discussion. Your idea of splitting the rename off is great, but you want that new PR to do both the rename to `_module.py` _and_ put back a new `module.py`. That PR diff will be large in GitHub, but that's fine. As long as the `git mv` and putting back the old name are two commits in a single PR, the git history is fine. We can merge that PR straight away, and master won't be broken.

After that, rebase this PR on master, and you have a small diff back here.

AnirudhDagar commented:
> This isn’t ideal, because we cannot merge PR 1 separately - that will break master, and the second PR may need more discussion. Your idea of splitting the rename off is great, but you want that new PR to do both the rename to _module.py and put back a new module.py. That PR diff will be large in GitHub, but that’s fine. As long as the git mv and putting back the old name are two commits in a single PR, the git history is fine. We can merge that PR straight away, and master won’t be broken.
After that, rebase this PR on master, and you have a small diff back here.

Please see the offline discussion below we had regarding the comment above.

Anirudh:
> Could you maybe elaborate on how PR 1 will break master?
Anyways, I feel I’ve overcomplicated things just to avoid the huge diff. If you think it is okay to make the merge with huge diff as long it still keeps git blame, I’m cool with taking that path too. :))

Ralf:

> Maybe CI is not broken, because we do not directly test(e.g. `filters.py`). There are still two reasons not to do this:
> 1. we do encourage downstream projects to test against master (just like we test NumPy master in Scipy CI)
> 2. we want to have the master branch releasable at any time. if CI is green but we have to "remember" that we cannot branch a release without restoring filters.py et al., that's less than ideal.

<hr/>

In a nutshell, to conclude, I agree it is a much better idea to follow what was initially suggested by Ralf as the above points make a lot of sense and the way I was pitching it earlier is not compliant with these points. I'll send a PR for achieving this now.

Thanks again everyone for the patience and useful feedback/suggestions. :))

AnirudhDagar committed:
STY: ndimage linting and doc fix

AnirudhDagar commented:
I've force pushed changes after rebasing on master (now that #14447 is merged). I request you to please re-review this. It should now address everything [mentioned](https://github.com/scipy/scipy/pull/14356#discussion_r672534127) by @grlee77 and also the [concern](https://github.com/scipy/scipy/pull/14356#discussion_r668071635) earlier raised by @peterbell10.

Let me know if there are any other changes/suggestions here. :)

grlee77 commented on scipy/ndimage/_multimethods.py:
```diff
@@ -0,0 +1,1065 @@
+           'white_tophat', 'zoom']
+
+
+_mark_non_coercible = functools.partial(Dispatchable, dispatch_type=np.ndarray,
```
This `_mark_non_coercible` is used for the `output` argument, but it cannot have a dispatch type of `np.ndarray` since we need to also support dtypes or strings that can represent dtypes.

e.g. output='f', output='float32', output=np.float32 are all acceptable in addition to arrays

I am not sure the best way to handle this. One option is just to define a class like

```Python
class ndimage_output:
    pass
```

and then set `dispatch_type=ndimage_output` and the `__ua_convert__` function could handle this using something like:

```Python
    if dispatch_type is ndimage_output:
        if isinstance(value, np.ndarray):
            return np.asarray(value)
        else:
            try:
                return np.dtype(str(value))
            except TypeError:
                return np.dtype(value)
```

Perhaps, @peterbell10 knows of a better way to handle this, though?

AnirudhDagar replied:
Done!

grlee77 commented on scipy/ndimage/_multimethods.py:
```diff
@@ -0,0 +1,1065 @@
+
+@_create_ndimage(_identity_arg_replacer)
+def generate_binary_structure(rank, connectivity):
+    return
```
```suggestion
    return ()
```
needs to return a tuple, not None

grlee77 commented on scipy/ndimage/_multimethods.py:
```diff
@@ -0,0 +1,1065 @@
+    """
+    def self_method(input, structure=None, iterations=1, mask=None,
+                    output=None, *args, **kwargs):
+        return (dispatchables[0], dispatchables[1], iterations, mask,
```
```suggestion
        return (dispatchables[0], dispatchables[1], iterations,
```

grlee77 commented on scipy/ndimage/_multimethods.py:
```diff
@@ -0,0 +1,1065 @@
+@_get_docs
+def binary_opening(input, structure=None, iterations=1, output=None,
+                   origin=0, mask=None, border_value=0, brute_force=False):
+    return input, structure, mask, _mark_non_coercible(output)
```
```suggestion
    return input, structure, _mark_non_coercible(output), mask
```

grlee77 commented on scipy/ndimage/_multimethods.py:
```diff
@@ -0,0 +1,1065 @@
+@_get_docs
+def binary_closing(input, structure=None, iterations=1, output=None,
+                   origin=0, mask=None, border_value=0, brute_force=False):
+    return input, structure, mask, _mark_non_coercible(output)
```
```suggestion
    return input, structure, _mark_non_coercible(output), mask
```

grlee77 commented on scipy/ndimage/_multimethods.py:
```diff
@@ -0,0 +1,1065 @@
+    """
+    uarray argument replacer for ``binary_opening`` and ``binary_closing``.
+    """
+    def self_method(input, structure=None, iterations=1,
+                    output=None, *args, **kwargs):
+        return (dispatchables[0], dispatchables[1],
+                iterations, dispatchables[2]) + args, kwargs
```
```suggestion
    def self_method(input, structure=None, iterations=1, output=None, origin=0,
                    mask=None, *args, **kwargs):
        return (dispatchables[0], dispatchables[1], iterations,
                dispatchables[2], origin, dispatchables[3]) + args, kwargs
```

grlee77 commented on scipy/ndimage/_multimethods.py:
```diff
@@ -0,0 +1,1065 @@
+    return self_method(*args, **kwargs)
+
+
+@_create_ndimage(_input_labels_index_replacer)
+@all_of_type(np.ndarray)
+@_get_docs
+def labeled_comprehension(input, labels, index, func, out_dtype,
+                          default, pass_positions=False):
+    return input, labels, index
```
```suggestion


def _labeled_comprehension_replacer(args, kwargs, dispatchables):
    """
    uarray argument replacer for signatures involving required
    ``input`` with ``labels`` and ``index`` kwargs.
    """
    def self_method(input, labels, index, func, out_dtype, *args,
                    **kwargs):
        return (dispatchables[0], dispatchables[1], dispatchables[2], func,
                dispatchables[3]) + args, kwargs

    return self_method(*args, **kwargs)


@_create_ndimage(_labeled_comprehension_replacer)
@all_of_type(np.ndarray)
@_get_docs
def labeled_comprehension(input, labels, index, func, out_dtype,
                          default, pass_positions=False):
    return input, labels, index, Dispatchable(out_dtype, np.dtype)
```

I think this should also dispatch on `out_dtype` as above?

AnirudhDagar replied:
Fixed now.

grlee77 reviewed (changes requested):
Thanks @AnirudhDagar! This is getting closer, but there are still a handful of remaining issues in the multimethods. I think there are various errors that aren't getting caught because we don't have any tests for a backend where `__ua_convert__` has been defined (this means the argument replacers are never called!)

I suggested fixes for several things that I found by running the existing test suite using a modified `_backend.py` where I temporarily set a `__ua_convert__` method as follows in `_ScipyImageBackend`

```Python

class ndimage_output():
    pass


class _ScipyImageBackend:
    """The default backend for ndimage calculations

    Notes
    -----
    We use the domain ``numpy.scipy`` rather than ``scipy`` because ``uarray``
    treats the domain as a hierarchy. This means the user can install a single
    backend for ``numpy`` and have it implement ``numpy.scipy.ndimage`` as
    well.

    .. versionadded:: 1.8.0
    """
    __ua_domain__ = "numpy.scipy.ndimage"


    def __ua_function__(method, args, kwargs):
        fn = getattr(_api, method.__name__, None)

        if fn is None:
            return NotImplemented
        return fn(*args, **kwargs)

    @ua.wrap_single_convertor
    def __ua_convert__(value, dispatch_type, coerce):
        if value is None:
            return None

        if dispatch_type is np.ndarray:
            if not coerce and not isinstance(value, np.ndarray):
                return NotImplemented

            return np.asarray(value)

        if dispatch_type is np.dtype:
            try:
                return np.dtype(str(value))
            except TypeError:
                return np.dtype(value)

        if dispatch_type is ndimage_output:
            if isinstance(value, np.ndarray):
                return np.asarray(value)
            else:
                try:
                    return np.dtype(str(value))
                except TypeError:
                    return np.dtype(value)

        return value
```

Similar to how `output` should not be dispatched simply as `np.ndarray` (see the suggestions), I think the `index` argument to various `ndimage.measurements` functions also needs some workaround since it can either be either a scalar integer or a NumPy array and the integer should **not** get converted to a 0-dim array scalar via `np.asarray`.

grlee77 commented on scipy/ndimage/_multimethods.py:
```diff
@@ -0,0 +1,1065 @@
+@_create_ndimage(_input_labels_index_replacer)
+@all_of_type(np.ndarray)
+@_get_docs
+def sum_labels(input, labels=None, index=None):
```
The index argument here and on most or all other methods in `_measurements.py` needs to also accept an integer, rather than always being coerced to an array.

AnirudhDagar replied:
I'm marking them as `_mark_index` and finally handling those also in `__uarray_convert__`. Does that sound right?

grlee77 commented:
@peterbell10, I have a couple of questions:

1.) Do you agree that we do not want to have a `__ua_convert__` in `_ScipyImageBackend` (we did not add one in `scipy.fft`'s backend)? Was that choice made there to maximize backwards compatibility, to avoid overhead of argument replacers or for another reason?

2.) What do you recommended for dispatching on the `output` argument (which may be an `np.ndarray` or a dtype). See comment: https://github.com/scipy/scipy/pull/14356#discussion_r676266260

Assuming we are not going to have `__ua_convert__` defined in `_ScipyImageBackend`, I think we still need need to run tests using a dummy backend something like the following so that we can make sure that all argument converters are working as expected. 

Note that I had to set `coerce=True` below because some test cases provide lists instead of arrays as the image input, but the multimethods dispatch `input` as `np.ndarray`.

```Python

import scipy._lib.uarray as ua
from scipy.ndimage._backend import _ScipyImageBackend, set_global_backend, ndimage_output


class _BackendWithConverter(_ScipyImageBackend):

    @ua.wrap_single_convertor
    def __ua_convert__(value, dispatch_type, coerce):
        if value is None:
            return None

        if dispatch_type is np.ndarray:
            if not coerce and not isinstance(value, np.ndarray):
                return NotImplemented

            return np.asarray(value)

        if dispatch_type is np.dtype:
            try:
                return np.dtype(str(value))
            except TypeError:
                return np.dtype(value)

        if dispatch_type is ndimage_output:
            if isinstance(value, np.ndarray):
                return np.asarray(value)
            else:
                try:
                    return np.dtype(str(value))
                except TypeError:
                    return np.dtype(value)

        return value

set_global_backend(_BackendWithConverter, coerce=True, try_last=True)
```

AnirudhDagar committed:
BUG: Fix positioning of arguments and dispatchables

AnirudhDagar commented:
Thanks @grlee77 for the detailed review and finding some of the issues. I've fixed the trivial ones. Let's wait for @peterbell10 to share his opinion on the `dispatch_type` and `__ua_convert__ ` issue. I'll then address those review comments according to the suggestions.

AnirudhDagar commented:
gentle ping @peterbell10, PTAL whenever you get some time. Thanks in advance :)

peterbell10 commented:
Sorry for the late response, the pings somehow missed my notifications completely. 

> 1.) Do you agree that we do not want to have a `__ua_convert__` in `_ScipyImageBackend` (we did not add one in `scipy.fft`'s backend)? Was that choice made there to maximize backwards compatibility, to avoid overhead of argument replacers or for another reason?

The answer is a bit of both. For backward compatibility the backend always coerces, so `__ua_convert__` is not necessary. Also, since it cuts out some of the protocol calls, it also reduces performance overhead. That said the overhead should have gone down since in the mean time so it's worth measuring.

In hindsight I think you can use `__ua_convert__` in a backward compatible way as well. You just have to set the SciPy backend with `coerce=True` so it will accept lists. Note that when the `scipy.fft` work started, global backends didn't have the `coerce` option.

> 2.) What do you recommended for dispatching on the output argument (which may be an `np.ndarray` or a `dtype`). See comment: https://github.com/scipy/scipy/pull/14356#discussion_r676266260

Ugh, that's unfortunate. I don't think `uarray` was designed for sum-types. I see that it's inconsistent between functions as well.

I can see two possible solutions:
1. define a new type to be either-or (as you suggest)
2. try to detect `dtype`-like arguments in the argument extractor and conditionally mark them as dispatchable.

The first option means every backend needs to be aware of this type and have special handling for it. However, the second won't work for non-NumPy dtypes which means I think we have to go with the first option.

> Assuming we are not going to have `__ua_convert__` defined in `_ScipyImageBackend`, I think we still need need to run tests using a dummy backend something like the following so that we can make sure that all argument converters are working as expected.

Agreed, that makes a lot of sense.   Argument replacers for complex signatures are fiddly and need to be tested.

AnirudhDagar committed:
BUG: Handle sum-types for output and index args

AnirudhDagar committed:
Merge branch 'master' into uarray_ndimage_backend

AnirudhDagar committed:
STY: Fix linting

AnirudhDagar commented:
Sorry for the slow updates during the last couple of weeks, was involved in other things. Hope to get this ready now.

I've made the changes to handle `output`/`index` args coming in multi-types. Also to validate the argument replacers, all the `ndimage` module tests now run with the `uarray` backend (using `__ua_convert__`) as suggested by @grlee77.

grlee77 reviewed (commented):
The new changes here look good. I guess we need to make a decision on whether to keep a `_BackendWithConverter` class where it is now in the test module or if we just move that `__ua_convert__` method to the SciPy backend itself. What do you think @rgommers, @peterbell10? I think the overhead is pretty negligible and it is mainly an issue of whether we missed some odd case that currently works, but is not handled correctly by `__ua_convert__` (although if there is any such a case, it is not used in our test suite).

We will need to add some documentation to help downstream libraries understand the conversions needed since it is a bit more complicated here than in the `scipy.fft` module. We can point to the `__ua_convert__` here as an example and mention the special handling of the `output` and `index` types.

AnirudhDagar committed:
ENH: Default backend with __ua_covert__ with coerce=True

grlee77 commented:
There are some real [failures in the doctests and tutorial examples](https://dev.azure.com/scipy-org/SciPy/_build/results?buildId=13788&view=logs&j=ede89041-5e8e-5201-ce5c-f68c2cbf02dd&t=bf6d2fae-c4b4-5aaf-e29f-d3092465da33&l=39). All seem to involve lists as input to the `input`, `weights` or `index` arguments, so apparently these are not getting coerced to arrays as intended.

AnirudhDagar committed:
ENH: Default backend args; coerce=True, try_last=True

AnirudhDagar commented:
@grlee77 I'm not sure how I can set `coerce=True` with `register_backend`. Any suggestions?
Thanks!

hameerabbasi commented:
`coerce=True` implies `only=True`, so it doesn't make sense for `register_backend` because then it would always take precedence.

AnirudhDagar commented:
Yeah, that is absolutely correct and makes sense. In that case, the current behaviour is probably not a good idea to have, since `register_backend` and `set_global_backend` will always be incompatible in behaviour.

hameerabbasi commented:
`register_backend` makes sense with `set_global_backend` with `try_last=True`, or with `coerce=False`, so all backends are tried.

AnirudhDagar commented:
The issue is that with the current implementation makes a case like [this](https://github.com/AnirudhDagar/scipy/blob/f174256be1318c0d792cdbeaeb22d4e67125e5d3/scipy/ndimage/_backend.py#L176-L187) problematic. This throws `uarray.BackendNotImplementedError: No selected backends had an implementation for this function.` as seen in the [CI](https://dev.azure.com/scipy-org/SciPy/_build/results?buildId=13834&view=logs&j=ede89041-5e8e-5201-ce5c-f68c2cbf02dd&t=bf6d2fae-c4b4-5aaf-e29f-d3092465da33&l=39)

hameerabbasi commented:
> The issue is that with the current implementation makes a case like [this](https://github.com/AnirudhDagar/scipy/blob/f174256be1318c0d792cdbeaeb22d4e67125e5d3/scipy/ndimage/_backend.py#L176-L187) problematic. This throws `uarray.BackendNotImplementedError: No selected backends had an implementation for this function.` as seen in the [CI](https://dev.azure.com/scipy-org/SciPy/_build/results?buildId=13834&view=logs&j=ede89041-5e8e-5201-ce5c-f68c2cbf02dd&t=bf6d2fae-c4b4-5aaf-e29f-d3092465da33&l=39)

I would argue this is not a valid use-case: A global backend should be able to handle everything, provided the types are correct. 😄

AnirudhDagar committed:
DOC: Update ndimage register_backend example

AnirudhDagar commented:
Thanks @hameerabbasi for clearing that out. I'll update that example slightly.

AnirudhDagar commented:
> We will need to add some documentation to help downstream libraries understand the conversions needed since it is a bit more complicated here than in the scipy.fft module. We can point to the __ua_convert__ here as an example and mention the special handling of the output and index types.

@rgommers any ideas where we can put such a small note in the docs? How about a small subheading like ["Code compatibility features"](https://docs.cupy.dev/en/stable/reference/scipy_fft.html#code-compatibility-features) (from CuPy docs)?

AnirudhDagar committed:
DOC: Add docs for __ua_convert__

AnirudhDagar committed:
Merge branch 'master' into uarray_ndimage_backend

AnirudhDagar commented:
From an offline discussion about this:
> Anirudh: @rgommers any ideas where we can put such a small note in the docs? How about a small subheading like "uarray Interoperabilty" (similar to what "Code Compatibility Features" is in CuPy docs)?
> Ralf: Let's put the uarray dev docs in https://github.com/scipy/scipy/tree/master/doc/source/dev/api-dev (link it into the doc build from https://github.com/scipy/scipy/blob/master/doc/source/dev/api-dev/api-dev-toc.rst)

I've updated the PR adding developer documentation regarding `__ua_convert__`.  You can see it rendered [here](https://36301-1460385-gh.circle-artifacts.com/0/html-scipyorg/dev/api-dev/uarray.html#uarray) which is linked through the `ndimage` _backend control_ documentation [here](https://36301-1460385-gh.circle-artifacts.com/0/html-scipyorg/reference/ndimage.html?highlight=ndimage#backend-control).

Hopefully, this should be ready for final review now. Apologies for the delays, I've been relocating from my home town in the past couple of weeks.

cc @peterbell10 @grlee77

grlee77 commented on doc/source/dev/api-dev/uarray.rst:
```diff
@@ -0,0 +1,50 @@
+in the ndimage module functions are : ``'output'`` and ``'index'``.
+
+* ``output``:
+  This is the output array which holds the output of a function.
+  It can take different types include a python ``str`` when output='f',
+  or ``output='float32'`` and dtypes like ``output=np.float32``.
```
"This is the output array" is a bit misleading since you explain below that it can also be other types. I tried to make it a bit clearer and added one set of missing double backticks.

```suggestion
  This can be either an actual array in which the function output will be
  stored or an indication of the data type for the output array. It can take
  different types, including a python ``str`` when ``output='f'`` or
  ``output='float32'`` and dtypes like ``output=np.float32``.
```

grlee77 commented on scipy/ndimage/_backend.py:
```diff
@@ -0,0 +1,270 @@
+            return np.asarray(value)
+
+        if dispatch_type is np.dtype:
+            try:
+                return np.dtype(str(value))
+            except TypeError:
+                return np.dtype(value)
```
Is there a reason this can't just be simplified to the following?:
```suggestion
            return np.dtype(value)
```

I see that this same try/except [is also in unumpy](https://github.com/Quansight-Labs/unumpy/blob/97dcc82e8f3343c2a66b5ee6d8081886652de1d0/unumpy/numpy_backend.py#L70-L74), but I am not sure why.  If you do change it here, please update the example in the `__ua_convert__` docs as well.

AnirudhDagar replied:
Not sure why it was there in the first place. Seems fine to me refactoring this. Done.

grlee77 commented on scipy/ndimage/_backend.py:
```diff
@@ -0,0 +1,270 @@
+            if isinstance(value, np.ndarray):
+                return np.asarray(value)
+            else:
+                try:
+                    return np.dtype(str(value))
+                except TypeError:
+                    return np.dtype(value)
```
same try/except as above

grlee77 reviewed (changes requested):
Thanks, @AnirudhDagar. The new doc page looks good to me. I had one text update suggestion and a question about simplifying the dtype handling in `__ua_convert__`.

AnirudhDagar committed:
MAINT: Remove redundant try except

AnirudhDagar commented:
@grlee77 I've addressed the comments, failing tests seem unrelated.

grlee77 commented on scipy/ndimage/_multimethods.py:
```diff
@@ -0,0 +1,1085 @@
+                                        coercible=False)
+_mark_index = functools.partial(Dispatchable,
+                                dispatch_type=ndimage_index,
+                                coercible=False)
```
```suggestion
                                coercible=True)
```
required if the suggested change to `__ua_convert__` is made

grlee77 reviewed (changes requested):
Thanks for the quick update @AnirudhDagar. 

I came up with a couple more small suggestions related to `__ua_convert__` (apologies for not catching these in the prior round of comments!)

AnirudhDagar committed:
MAINT: Update __ua_convert__

grlee77 reviewed (approved):
Thanks, looks good to me.

AnirudhDagar commented:
Thanks, @grlee77 for being patient with this PR. Let's wait for @peterbell10 to review this.

Ps. If you'd squash merge then the PR is fine, otherwise, let me know I can do an interactive rebase to make the commit history cleaner before a merge commit.

peterbell10 commented on scipy/ndimage/_backend.py:
```diff
@@ -0,0 +1,266 @@
+            if isinstance(value, np.ndarray):
+                return value
+            else:
+                return np.dtype(value)
```
This doesn't allow non-numpy array types.

AnirudhDagar replied:
Should be fixed with this.
```python
        if dispatch_type is ndimage_output:
            if not isinstance(value, (np.ndarray, Type, np.dtype, str)):
                return NotImplemented
            elif isinstance(value, np.ndarray):
                return value
            else:
                return np.dtype(value)
```

AnirudhDagar replied:
I don't think this is the right way to go. Assume an array object that supports the `__array_interface__` protocol, it will then be coerced to numpy array which is not desirable. The output is supposed to be non-coercible.

peterbell10 replied:
No, I don't think `__array_interface__` should be used. It might be possible that some forms of coercion will work out, but I would expect that to lead to bugs where the output isn't actually written in-place. Practically, I also see `isinstance(output, np.ndarray)` throughout the ndimage code so I wouldn't expect it to work with `return value` either.

peterbell10 commented on scipy/ndimage/_backend.py:
```diff
@@ -0,0 +1,266 @@
+    return ua.skip_backend(backend)
+
+
+set_global_backend('scipy')
```
Are there no tests passing lists? I would expect that to fail without setting `coerce=True`.
```suggestion
set_global_backend('scipy', coerce=True)
```

AnirudhDagar replied:
`coerce=True` is a default in ndimage `set_global_backend`. See [LOC100](https://github.com/scipy/scipy/blob/7c4ce13a58a12b01fa0a3b1fe0aeb09db92df8f4/scipy/ndimage/_backend.py#L100) in `_backend.py`.

AnirudhDagar replied:
There are definitely a few tests that provide lists as input.

peterbell10 commented on scipy/ndimage/_multimethods.py:
```diff
@@ -0,0 +1,1085 @@
+           'white_tophat', 'zoom']
+
+
+_mark_non_coercible = functools.partial(Dispatchable,
```
`_mark_output` would be more accurate.

AnirudhDagar replied:
Yes, this name makes more sense. I will update that.

jakirkham commented:
cc @GenevieveBuckley (in case you have thoughts on this 🙂)

AnirudhDagar committed:
MAINT: Refactor _mark_non_coercible -> _mark_output

AnirudhDagar committed:
MAINT: raise if not numpy array or dtype for ndimage_output

AnirudhDagar committed:
DOC: Update uarray dev docs for ndimage_output

AnirudhDagar committed:
Merge branch 'master' into uarray_ndimage_backend

GenevieveBuckley commented:
Thanks for pinging me, @jakirkham 

> Let me know if it is necessary to include this type of test for ndimage as well.

I think this seems important. Apologies if this is something that's already been discussed/resolved, there's a going on here in the thread.

AnirudhDagar commented:
Hi @GenevieveBuckley,
Yeah, we are in fact running the whole `ndimage` test suite on `uarray`. See this [comment](https://github.com/scipy/scipy/pull/14356#pullrequestreview-714413987). Sorry for not updating the PR description (will do that now).

peterbell10 commented on scipy/ndimage/_backend.py:
```diff
@@ -0,0 +1,269 @@
+            return np.dtype(value)
+
+        if dispatch_type is ndimage_output:
+            if not isinstance(value, (np.ndarray, Type, np.dtype, str)):
```
Is there any particular reason to check `typing.Type` instead of the builtin `type`?

AnirudhDagar replied:
Yeah, I should have probably just used type, made it complicated for no reason.
Sorry. Fixed now!

AnirudhDagar committed:
MAINT: Use type instead of typing.Type

AnirudhDagar commented:
Thanks, @grlee77 @peterbell10 for reviewing this multiple times, suggesting changes and overall improving the work here. I guess we are finally there. Let me know if we'd like to keep the commit history (in that case I can do an interactive rebase and improve if required) or squash the commits on merge.

cc @rgommers

AnirudhDagar commented:
I don't have any strong opinions on this. Let's do a squash merge if that's ideal.

rgommers reviewed (changes requested):
This looks excellent, thanks a lot @AnirudhDagar for the hard work, and @peterbell10 and @grlee77 for the thorough reviews!

I'm just adding a blocking review for the sole reason that we should not merge this just yet. For the same reason as discussed in https://github.com/scipy/scipy/pull/14407#discussion_r674622670: we need to (a) sketch the big picture of where this uarray interop work is going, and (b) have a discussion about it on the mailing list.

@AnirudhDagar wrote a very nice blog post that will help illustrate where we are heading: https://labs.quansight.org/blog/2021/10/array-libraries-interoperability/. A second blog post is in the making with more of a high-level technical overview - that should be ready within a few days (*so please do not start the discussion now/here*). Once that is published, we will have that discussion, and once it has settled down we can (presumably) merge this PR.

IvanYashchuk commented on scipy/ndimage/_backend.py:
```diff
@@ -0,0 +1,268 @@
+
+    .. versionadded:: 1.8.0
+    """
+    __ua_domain__ = "numpy.scipy.ndimage"
```
I understand the intention and I know this is done the same way for `scipy.fft`, but it's weird to call the hypothetical top-level backend `numpy`.
uarray supports a list of domains that can be used when implementing a single backend that supports multiple domains https://github.com/Quansight-Labs/uarray/pull/243 and it's just easier to understand the domain hierarchy when there's no `"numpy."` at the beginning of the string.
@hameerabbasi, @peterbell10 what do you think, is it really necessary to have `numpy` here?

tylerjereddy commented:
@rgommers Where does the team stand on the overall design decisions/blocking comment above? Probably bet to bump the milestone before we branch ~later this week?

rgommers commented:
Yes, we're going to have to bump this unfortunately - not enough bandwidth to push these discussions forward right now. Moving to the 1.10.0 milestone.

j-bowhay commented:
Closing as we now have array api delegation.
